core/src/main/scala/zio/config/ConfigDescriptor.scala

```diff
@@ -170,15 +171,66 @@ object ConfigDescriptor {
   val float: ConfigDescriptor[String, String, Float] =
     ConfigDescriptor.Source(ConfigSource.empty, PropertyType.FloatType) ?? "value of type float"
 
+  def first[K, V, A](path: K)(desc: ConfigDescriptor[K, V, A]) =
+    list(path)(desc)
+      .xmapEither[A](_.headOption.fold[Either[String, A]](Left("Element is missing"))(Right(_)), v => Right(v :: Nil))
+
   def float(path: String): ConfigDescriptor[String, String, Float] = nested(path)(float)
 
   val int: ConfigDescriptor[String, String, Int] =
     ConfigDescriptor.Source(ConfigSource.empty, PropertyType.IntType) ?? "value of type int"
 
   def int(path: String): ConfigDescriptor[String, String, Int] = nested(path)(int)
 
-  def list[K, V, A](desc: ConfigDescriptor[K, V, A]): ConfigDescriptor[K, V, List[A]] =
-    ConfigDescriptor.Sequence(desc)
+  /**
+   * Leaks out path in simple cases, i.e.
+   * `list(string(path))` == `list(path)(string)` == `nested(path)(listOrSingle(string))`
+   *
+   * `nested("a")(list(string("b")))` describes configuration `{a: { b: ["s1", "s2"] }}`
+   *
+   * Allows scalar value instead of list
+   * */
+  def list[K, V, A](desc: ConfigDescriptor[K, V, A]): ConfigDescriptor[K, V, List[A]] = {
+    def extractPath(cfg: ConfigDescriptor[K, V, A]): Option[(K, ConfigDescriptor[K, V, A])] = cfg match {
+      case Describe(config, message) =>
+        extractPath(config).map { case (path, inner) => (path, Describe(inner, message)) }
+      case Nested(path, config) => Some((path, config))
+      case _                    => None
+    }
+
+    extractPath(desc) match {
+      case Some((path, inner)) => list(path)(inner)
+      case None                => listOrSingle(desc)
+    }
+  }
+
+  /**
+   * Keeps inner description intact.
+   *
+   * Allows scalar value instead of list
+   * */
+  def list[K, V, A](path: K)(desc: ConfigDescriptor[K, V, A]): ConfigDescriptor[K, V, List[A]] =
+    nested(path)(listOrSingle(desc))
+
+  /**
+   * Keeps inner description intact.
+   *
+   * `nested("a")(listOrSingle(string("b")))` describes configuration `{a: [{b: "s1"}, {b: "s2"}]}`
+   *
+   * Allows scalar value instead of list
+   * */
+  def listOrSingle[K, V, A](desc: ConfigDescriptor[K, V, A]): ConfigDescriptor[K, V, List[A]] =
```

Comment:
By consistency with previous conventions, this one should be called `list` only, because the above `list` with `path` just delegates to this one, calling `nested` to wrap it into the specified field of a map.

```suggestion
  def list[K, V, A](desc: ConfigDescriptor[K, V, A]): ConfigDescriptor[K, V, List[A]] =
```

Reply:
@jdegoes unfortunately to preserve current behavior of `list(string(path))` we need `list` to leak out nested path.

Reply:
Ah, I see. Let's just break backward compatibility here. The new way is much more precise and consistent.

Reply:
done